solutions/cfgym-101102D.mdx

```diff
@@ -0,0 +1,259 @@
+---
+id: cfgym-101102D
+source: CF
+title: Rectangles
+author: Kevin Sheng
+---
+
+Let's first simplify this problem, and then apply the solution of that problem to this one.
+
+Say we had a matrix of all `0`'s and `1`'s. How many submatrices of all `1`'s are there?
+
+We first define a 2D array, $\texttt{longest\_consec}$, that contains for each cell how many `1`'s there are to the right.  
+For example, this matrix:
+
+<table>
+	<tr>
+		<td>1</td>
+		<td>0</td>
+		<td>1</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>1</td>
+		<td>1</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>0</td>
+		<td>0</td>
+		<td>1</td>
+	</tr>
+</table>
+
+Would have the following $\texttt{longest\_consec}$:
+
+<table>
+	<tr>
+		<td>1</td>
+		<td>0</td>
+		<td>2</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>3</td>
+		<td>2</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>0</td>
+		<td>0</td>
+		<td>1</td>
+	</tr>
+</table>
+
+Now, we go through each column and then for each row in reverse,
+find out for each cell the number of all `1`'s submatrices where the cell is the upper left corner.
+
+How do we do this?  
```

Comment:
not necessary

Reply:
tbh i feel it makes the editorial flow better

Reply:
sure